File: gen_coverage/benchmark.py
import time
import shutil
import subprocess
import json
import math
from pathlib import Path
from concurrent.futures import ProcessPoolExecutor, as_completed

from .gcov import get_gcov_env, process_prefix, get_prefix, get_prefix_files, combine_reports, symlink_gcno_files
from .utils import sample_files
from . import MIN_JOB_SIZE, PROGRESS_MANAGER

def process_file(file, cmd_arg, build_dir, batch_id=None, use_prefix=False, verbose=False):
    """Process a single file with the given command."""
    res = f"| File: {file}\n"
    start_time = time.time()
    bid_msg = "" if batch_id is None else f" in batch {batch_id}"

    try:
        if use_prefix:
            result = subprocess.run(cmd_arg + [file], env=get_gcov_env(file), check=True, capture_output=True, text=True,)
        else:
            result = subprocess.run(cmd_arg + [file], check=True, capture_output=True, text=True,)
        sout = result.stdout[:-1]
    except subprocess.CalledProcessError as e:
        if e.returncode == -6:
            sout = "timeout"
        else:
            res += f"Error processing file {file}: {e}\n"
            sout = f"crash (returncode: {e.returncode})"

    res += f"{sout}\n"
    duration = (time.time() - start_time) * 1000  # Convert to milliseconds
    res += f"-> Execution Time: {duration:.2f} ms\n"
    if verbose: print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Execution of /.../{'/'.join(Path(file).parts[-5:])}{bid_msg}:\n{sout}")

    start_time = time.time()

    prefix = get_prefix(file)
    files = get_prefix_files(prefix)
    symlink_gcno_files(build_dir, prefix)
    files_report = process_prefix(prefix, files)

    # Delete the folder to keep storage available
    shutil.rmtree(prefix)

    duration = (time.time() - start_time) * 1000  # Convert to milliseconds
    res += f"-> Processing Time: {duration:.2f} ms\n"
    if verbose: print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Processed prefix for /.../{'/'.join(Path(file).parts[-5:])}{bid_msg}")

    return (res, files_report)

def process_file_batch(file_batch, cmd_arg, build_dir, batch_id=None, use_prefix=False, verbose=False):
    log = ""
    report = { "sources": {} }
    fb_len = len(file_batch)
    for (i, file) in enumerate(file_batch):
        (flog, freport) = process_file(file, cmd_arg, build_dir, batch_id, use_prefix=use_prefix)
        log += flog
        combine_reports(report, freport, exec_one=False)
        if verbose: print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Combined intermediate results for batch {batch_id}")
        if i != fb_len and i % 5 == 4: 
            print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Batch {batch_id}: Processed file {i + 1} of {fb_len}")

    return (log, report, fb_len)


def run_benchmark(sample_size, benchmark_dir, job_size, cmd_arg, bname, build_dir, use_prefix=False, verbose=False):
    """Run the benchmark on sampled files."""
    report = { "sources": {} }
    print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Retrieving files to be benchmarked... (file count: {sample_size})")
    files = sample_files(sample_size, benchmark_dir)
    log_path = Path(f"{bname}.log")
    log_file = open(log_path, 'w')

    log_file.write(f"Running benchmark on {sample_size} test files in {benchmark_dir}\n")
    log_file.write("\n-------------------------------------\n")
    
    print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Starting benchmarks...")

    start_time = time.time()

    # Run commands either in parallel or sequentially
    if job_size > 1:
        batch_size = max(job_size, math.ceil(len(files) / MIN_JOB_SIZE))
        file_batches = [files[i::batch_size] for i in range(batch_size)]
        pbar = PROGRESS_MANAGER.counter(total=len(file_batches), desc='Processing batches', unit='batches')

        with ProcessPoolExecutor(max_workers=job_size) as executor:
            print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Processing of {len(file_batches)} batches (exp. batch_size: {len((file_batches[0:1] or [])[0])}) in {job_size} processes starts now...")
            futures = { executor.submit(process_file_batch, batch, cmd_arg, build_dir, batch_id, use_prefix): batch_id 
                        for batch_id, batch in enumerate(file_batches)}
            future_len = len(futures)
            for i, future in enumerate(as_completed(futures)):
                (log, files_report, batch_size) = future.result()
                log_file.write(log + '\n')
                combine_reports(report, files_report, exec_one=False)
                pbar.update()
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] Finished batch {i + 1} of {future_len} (batch_size: {batch_size})")

    else:
        pbar = PROGRESS_MANAGER.counter(total=len(files), desc='Processing files', unit='files')
        for file in files:
            (log, files_report) = process_file(file, cmd_arg, build_dir, None, use_prefix=use_prefix)
            log_file.write(log + '\n')
            combine_reports(report, files_report, exec_one=False)
            pbar.update()

    with open(f"{bname}_coverage.json", "w") as f:
        json.dump(report, f)

    duration = (time.time() - start_time)
    msg = f"=> Total Benchmark Runtime: {duration:.2f}s"
    log_file.write(msg + '\n')
    print(f"\n[{time.strftime('%Y-%m-%d %H:%M:%S')}] {msg}")
    if job_size > 1:
        msg = f"=> Avg. Benchmark Runtimes: {(duration / len(files)):.2f} s/file {duration / len(file_batches):.2f} s/batch"
        log_file.write(msg + '\n')
        print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {msg}")
    else:
        msg = f"=> Avg. Runtime: {(duration / len(files)):.2f} s/file"
        log_file.write(msg + '\n')
        print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {msg}")


    log_file.flush()
    log_file.close()

File: rust/src/benchmark.rs
use rayon::prelude::*;
use serde_json::Value;
use std::fs::{self, File};
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use crate::args::ARGS;
use crate::gcov::{get_gcov_env, get_prefix, get_prefix_files, process_prefix, symlink_gcno_files};
use crate::utils::{combine_reports, sample_files};

pub fn process_file(
    file: &str,
    batch_id: Option<usize>,
) -> (String, Value) {
    let mut res = format!("| File: {}\n", file);
    let start_time = Instant::now();
    let bid_msg = if let Some(id) = batch_id {
        format!(" in batch {}", id)
    } else {
        String::new()
    };
    let cmd = Path::join(&ARGS.build_dir, "bin/cvc5");

    let output = if ARGS.individual_prefixes {
        let file_path = PathBuf::from(file);
        let env = get_gcov_env(&file_path);
        Command::new(cmd)
            .args([&ARGS.cvc5_args])
            .arg(file)
            .envs(&env)
            .output()
    } else {
        Command::new(cmd).args([&ARGS.cvc5_args]).arg(file).output()
    };

    match output {
        Ok(result) => {
            let sout = String::from_utf8_lossy(&result.stdout);
            res += &sout;
        }
        Err(e) => {
            res += &format!("Error processing file {}: {}\n", file, e);
        }
    };

    let duration = start_time.elapsed().as_millis();
    res += &format!("-> Execution Time: {} ms\n", duration);

    if ARGS.verbose {
        println!(
            "[{}] Execution of /{}/{}{}:\n{}",
            chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
            ARGS.build_dir.display().to_string(),
            file,
            bid_msg,
            res
        );
    }

    // Processing prefix
    let start_time = Instant::now();
    let prefix = get_prefix(&PathBuf::from(file));
    let files = get_prefix_files(&prefix).unwrap();
    symlink_gcno_files(&prefix).unwrap();
    let files_report = process_prefix(
        &prefix,
        &files,
    )
    .unwrap();

    fs::remove_dir_all(&prefix).unwrap();

    let duration = start_time.elapsed().as_millis();
    res += &format!("-> Processing Time: {} ms\n", duration);

    if ARGS.verbose {
        println!(
            "[{}] Processed prefix for /{}/{}{}",
            chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
            ARGS.build_dir.display().to_string(),
            file,
            bid_msg
        );
    }

    (res, files_report)
}

pub fn process_file_batch(
    file_batch: Vec<String>,
    batch_id: Option<usize>,
) -> (String, Value) {
    let mut log = String::new();
    let mut report = serde_json::json!({"sources": {}});

    for (i, file) in file_batch.iter().enumerate() {
        let (flog, freport) = process_file(file, batch_id);
        log += &flog;
        combine_reports(&mut report, &freport, false);
        if ARGS.verbose {
            println!(
                "[{}] Combined intermediate results for batch {:?}",
                chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
                batch_id
            );
        }
        if i % 5 == 4 {
            println!(
                "[{}] Batch {:?}: Processed file {} of {}",
                chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
                batch_id,
                i + 1,
                file_batch.len()
            );
        }
    }

    (log, report)
}

pub fn run_benchmark(
    sample_size: &str,
    bname: &Path,
) -> io::Result<()> {
    let mut report = serde_json::json!({"sources": {}});

    println!(
        "[{}] Retrieving files to be benchmarked... (file count: {})",
        chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
        sample_size
    );
    let files = sample_files(sample_size);
    let log_path = bname.with_extension("log");
    let log_file = Arc::new(Mutex::new(File::create(log_path)?));

    writeln!(
        log_file.lock().unwrap(),
        "Running benchmark on {} test files in {}\n",
        sample_size,
        ARGS.benchmark_dir.display().to_string()
    )?;
    writeln!(
        log_file.lock().unwrap(),
        "\n-------------------------------------\n"
    )?;

    println!(
        "[{}] Starting benchmarks...",
        chrono::Local::now().format("%Y-%m-%d %H:%M:%S")
    );

    let start_time = Instant::now();

    if ARGS.job_size > 1 {
        let batch_size = std::cmp::max(ARGS.job_size, (files.len() + ARGS.job_size - 1) / ARGS.job_size);
        let file_batches: Vec<Vec<String>> = files
            .chunks(batch_size)
            .map(|chunk| chunk.to_vec())
            .collect();

        println!(
            "[{}] Processing {} batches in {} parallel jobs...",
            chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
            file_batches.len(),
            ARGS.job_size
        );

        let report = Arc::new(Mutex::new(report));
        file_batches.par_iter().for_each(|batch| {
            let (log, batch_report) = process_file_batch(
                batch.to_vec(),
                None,
            );
            let mut report_lock = report.lock().unwrap();
            combine_reports(&mut *report_lock, &batch_report, false);
            log_file.lock().unwrap().write_all(log.as_bytes()).unwrap();
        });
    } else {
        for file in files {
            let (log, files_report) =
                process_file(&file, None);
            log_file.lock().unwrap().write_all(log.as_bytes())?;
            combine_reports(&mut report, &files_report, false);
        }
    }

    let duration = start_time.elapsed().as_secs_f64();
    let msg = format!("=> Total Benchmark Runtime: {:.2}s", duration);
    writeln!(log_file.lock().unwrap(), "{}", msg)?;
    println!(
        "\n[{}] {}",
        chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
        msg
    );

    Ok(())
}
